library/footer.cpp

```diff
@@ -0,0 +1,289 @@
+ * It will checks only chunks that intersect @wc->offset and @wc->size.
+ *
+ * Returns negative error value or zero on success.
+ */
+static int eblob_verify_crc32(struct eblob_backend *b, struct eblob_key *key, struct eblob_write_control *wc) {
+	int err = 0;
+	uint64_t footers_offset = 0,
+	         footers_size = 0;
+
+	std::vector<uint32_t> calc_footers, check_footers;
+
+	err = eblob_chunked_crc32(b, key, wc, wc->offset, wc->size, calc_footers, footers_offset);
+	if (err) {
+		eblob_log(b->cfg.log, EBLOB_LOG_ERROR, "blob i%d: %s: %s: eblob_chunked_crc32: failed: fd: %d, size: %"PRIu64
+		          ", offset: %" PRIu64 "\n",
+		          wc->bctl->index, eblob_dump_id(key->id), __func__, wc->data_fd, footers_size, footers_offset);
+		return err;
+	}
+
+	footers_size = calc_footers.size() * sizeof(calc_footers.front());
+
+	check_footers.resize(calc_footers.size(), 0);
+	err = __eblob_read_ll(wc->data_fd, check_footers.data(), footers_size, footers_offset);
+	if (err) {
+		eblob_log(b->cfg.log, EBLOB_LOG_ERROR, "blob i%d: %s: %s: failed to read footer: fd: %d, size: %"PRIu64
+		          ", offset: %" PRIu64 "\n",
+		          wc->index, eblob_dump_id(key->id), __func__, wc->data_fd, footers_size, footers_offset);
+		return err;
+	}
+
+	if (memcmp(calc_footers.data(), check_footers.data(), footers_size)) {
+		eblob_log(b->cfg.log, EBLOB_LOG_DEBUG, "blob i%d: %s: %s: checksum mismatch: footers_size: %" PRIu64
+		          ", footers_count: %" PRIu64"\n",
+		          wc->index, eblob_dump_id(key->id), __func__, footers_size, calc_footers.size());
+		return -EILSEQ;
+	}
+
+	eblob_log(b->cfg.log, EBLOB_LOG_NOTICE, "blob: i%d: %s: %s: checksum verified\n",
+	          wc->index, eblob_dump_id(key->id), __func__);
+
+	return 0;
+}
+
+int eblob_verify_checksum(struct eblob_backend *b, struct eblob_key *key, struct eblob_write_control *wc) {
+	if (b->cfg.blob_flags & EBLOB_NO_FOOTER ||
+	    wc->flags & BLOB_DISK_CTL_NOCSUM)
+		return 0;
+
+	if (wc->total_size < wc->total_data_size)
+		return -EINVAL;
+
+	FORMATTED(HANDY_TIMER_SCOPE, ("eblob.%u.verify_checksum", b->cfg.stat_id));
+
+	int err;
+	if (wc->flags & BLOB_DISK_CTL_CHUNKED_CRC32)
+		err = eblob_verify_crc32(b, key, wc);
+	else
+		err = eblob_verify_sha512(b, key, wc);
+	return err;
+}
+
+int eblob_commit_footer(struct eblob_backend *b, struct eblob_key *key, struct eblob_write_control *wc) {
+	/*
+	 * skip footer committing if eblob is configured with EBLOB_NO_FOOTER flag or
+	 * the record should not be checksummed
+	 */
+	if (b->cfg.blob_flags & EBLOB_NO_FOOTER ||
+	    wc->flags & BLOB_DISK_CTL_NOCSUM)
+		return 0;
+
+	FORMATTED(HANDY_TIMER_SCOPE, ("eblob.%u.write.commit.footer", b->cfg.stat_id));
+
+	int err;
+	std::vector<uint32_t> checksums;
+	uint64_t checksums_offset;
+
+	/* calculates chunked crc32 of whole record's data */
+	err = eblob_chunked_crc32(b, key, wc, 0, wc->total_data_size, checksums, checksums_offset);
```

Comment:
We should write that checksum is calculated and written after the whole file has been written, i.e. at commit time and not during partial writes even if chunks are being completely written earlier